File: platform/windows/MihomoMeter.Windows.Tests/SharedCoreProxyTypeShadowTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using MihomoMeter.Windows.Core.Application;
using MihomoMeter.Windows.Core.Domain;

namespace MihomoMeter.Windows.Tests;

[TestClass]
public sealed class SharedCoreProxyTypeShadowTests
{
    private static readonly ProxyClassification Proxy = new(TrafficCategory.Proxy);
    private static readonly ProxyClassification Direct = new(TrafficCategory.Direct);
    private static readonly ProxyClassification Reject = new(TrafficCategory.Reject);
    private static readonly ProxyClassification Unknown = new(
        TrafficCategory.Unknown,
        UnknownTrafficReason.AmbiguousProxyType);

    [TestMethod]
    public void RouterMatchesEveryStableSharedClassificationWithoutChangingNativeResult()
    {
        (string RawType, ProxyClassification Native, SharedProxyTypeClassification Shared)[] cases =
        [
            ("Vmess", Proxy, SharedProxyTypeClassification.Proxy),
            ("Direct", Direct, SharedProxyTypeClassification.Direct),
            ("Reject", Reject, SharedProxyTypeClassification.Reject),
        ];

        foreach (var testCase in cases)
        {
            var result = SharedCoreProxyTypeRouter.Route(
                testCase.RawType,
                testCase.Native,
                _ => testCase.Shared);

            Assert.AreEqual(testCase.Native, result.Classification);
            Assert.AreEqual(SharedCoreProxyTypeRouteSource.SharedShadow, result.Source);
            Assert.AreEqual(SharedCoreProxyTypeRouteStatus.Matched, result.Status);
        }
    }

    [TestMethod]
    public void RouterKeepsNativeResultForUnrecognizedMismatchAndAdapterFailures()
    {
        var unrecognized = Route(SharedProxyTypeClassification.Unrecognized, Unknown);
        var mismatch = Route(SharedProxyTypeClassification.Direct, Proxy);
        (SharedProxyTypeAdapterFailure Failure, SharedCoreProxyTypeRouteStatus Status)[] failures =
        [
            (SharedProxyTypeAdapterFailure.UnsupportedAbiVersion,
                SharedCoreProxyTypeRouteStatus.AbiMismatch),
            (SharedProxyTypeAdapterFailure.NativeCallFailed,
                SharedCoreProxyTypeRouteStatus.NativeCallFailed),
            (SharedProxyTypeAdapterFailure.UnsupportedProxyTypeInput,
                SharedCoreProxyTypeRouteStatus.UnsupportedInput),
            (SharedProxyTypeAdapterFailure.ProxyTypeInputTooLong,
                SharedCoreProxyTypeRouteStatus.InputTooLong),
            (SharedProxyTypeAdapterFailure.UnsupportedProxyTypeCategory,
                SharedCoreProxyTypeRouteStatus.UnexpectedResult),
        ];

        Assert.AreEqual(Unknown, unrecognized.Classification);
        Assert.AreEqual(SharedCoreProxyTypeRouteSource.NativeFallback, unrecognized.Source);
        Assert.AreEqual(SharedCoreProxyTypeRouteStatus.Unrecognized, unrecognized.Status);
        Assert.AreEqual(Proxy, mismatch.Classification);
        Assert.AreEqual(SharedCoreProxyTypeRouteStatus.Mismatch, mismatch.Status);

        foreach (var failure in failures)
        {
            var result = SharedCoreProxyTypeRouter.Route(
                "Synthetic",
                Proxy,
                _ => throw new SharedProxyTypeAdapterException(
                    failure.Failure,
                    "synthetic"));
            Assert.AreEqual(Proxy, result.Classification);
            Assert.AreEqual(SharedCoreProxyTypeRouteSource.NativeFallback, result.Source);
            Assert.AreEqual(failure.Status, result.Status);
        }

        var unknownFailure = SharedCoreProxyTypeRouter.Route(
            "Synthetic",
            Proxy,
            _ => throw new ArgumentException("synthetic"));
        Assert.AreEqual(Proxy, unknownFailure.Classification);
        Assert.AreEqual(SharedCoreProxyTypeRouteStatus.UnknownFailure, unknownFailure.Status);
    }

    [TestMethod]
    public void ShadowDeduplicatesSourceAndStatusAndIgnoresReporterFailure()
    {
        var observations = new List<SharedCoreProxyTypeShadowObservation>();
        SharedCoreProxyTypeShadow.ConfigureReporter(observation =>
        {
            observations.Add(observation);
            throw new InvalidOperationException("synthetic");
        });
        try
        {
            for (var index = 0; index < 2; index += 1)
            {
                Assert.AreEqual(
                    Proxy,
                    SharedCoreProxyTypeShadow.Observe(
                        "Vmess",
                        Proxy,
                        _ => SharedProxyTypeClassification.Proxy));
            }
            Assert.AreEqual(
                Unknown,
                SharedCoreProxyTypeShadow.Observe(
                    "Selector",
                    Unknown,
                    _ => SharedProxyTypeClassification.Unrecognized));

            CollectionAssert.AreEqual(
                new[]
                {
                    new SharedCoreProxyTypeShadowObservation(
                        SharedCoreProxyTypeRouteSource.SharedShadow,
                        SharedCoreProxyTypeRouteStatus.Matched),
                    new SharedCoreProxyTypeShadowObservation(
                        SharedCoreProxyTypeRouteSource.NativeFallback,
                        SharedCoreProxyTypeRouteStatus.Unrecognized),
                },
                observations);
        }
        finally
        {
            SharedCoreProxyTypeShadow.ConfigureReporter(null);
        }
    }

    [TestMethod]
    public void ObservationGateKeysBySourceAndStatus()
    {
        var gate = new SharedCoreProxyTypeShadowObservationGate();
        var matched = new SharedCoreProxyTypeShadowObservation(
            SharedCoreProxyTypeRouteSource.SharedShadow,
            SharedCoreProxyTypeRouteStatus.Matched);
        var fallback = new SharedCoreProxyTypeShadowObservation(
            SharedCoreProxyTypeRouteSource.NativeFallback,
            SharedCoreProxyTypeRouteStatus.Matched);

        Assert.IsTrue(gate.ShouldReport(matched));
        Assert.IsFalse(gate.ShouldReport(matched));
        Assert.IsTrue(gate.ShouldReport(fallback));
        gate.Reset();
        Assert.IsTrue(gate.ShouldReport(matched));
    }

    private static SharedCoreProxyTypeRouteResult Route(
        SharedProxyTypeClassification sharedClassification,
        ProxyClassification nativeClassification)
    {
        return SharedCoreProxyTypeRouter.Route(
            "Synthetic",
            nativeClassification,
            _ => sharedClassification);
    }
}

File: Tests/SharedCoreProxyTypeShadowTests.swift
import XCTest
import os

@testable import MihomoMeter

final class SharedCoreProxyTypeShadowTests: XCTestCase {
  private let proxy = ProxyClassification(category: .proxy, unknownReason: nil)
  private let direct = ProxyClassification(category: .direct, unknownReason: nil)
  private let reject = ProxyClassification(category: .reject, unknownReason: nil)
  private let unknown = ProxyClassification(
    category: .unknown,
    unknownReason: .ambiguousProxyType
  )

  func testRouterMatchesEveryStableSharedClassificationWithoutChangingNativeResult() {
    let cases: [(String, ProxyClassification, SharedProxyTypeClassification)] = [
      ("Vmess", proxy, .proxy),
      ("Direct", direct, .direct),
      ("Reject", reject, .reject),
    ]

    for (rawType, nativeClassification, sharedClassification) in cases {
      let result = SharedCoreProxyTypeRouter.route(
        rawType: rawType,
        nativeClassification: nativeClassification,
        classifyProxyType: { _ in sharedClassification }
      )

      XCTAssertEqual(result.classification, nativeClassification)
      XCTAssertEqual(result.source, .sharedShadow)
      XCTAssertEqual(result.status, .matched)
    }
  }

  func testRouterKeepsNativeResultForUnrecognizedMismatchAndAdapterFailures() {
    let unrecognized = route(sharedResult: .unrecognized, nativeClassification: unknown)
    let mismatch = route(sharedResult: .direct, nativeClassification: proxy)
    let failures: [(SharedProxyTypeAdapterError, SharedCoreProxyTypeRouteStatus)] = [
      (.unsupportedABIVersion(2), .abiMismatch),
      (.nativeCallFailed(-1), .nativeCallFailed),
      (.unsupportedProxyTypeInput, .unsupportedInput),
      (.proxyTypeInputTooLong, .inputTooLong),
      (.unsupportedProxyTypeCategory(99), .unexpectedResult),
    ]

    XCTAssertEqual(unrecognized.classification, unknown)
    XCTAssertEqual(unrecognized.source, .nativeFallback)
    XCTAssertEqual(unrecognized.status, .unrecognized)
    XCTAssertEqual(mismatch.classification, proxy)
    XCTAssertEqual(mismatch.source, .nativeFallback)
    XCTAssertEqual(mismatch.status, .mismatch)

    for (error, expectedStatus) in failures {
      let result = SharedCoreProxyTypeRouter.route(
        rawType: "Synthetic",
        nativeClassification: proxy,
        classifyProxyType: { _ in throw error }
      )
      XCTAssertEqual(result.classification, proxy)
      XCTAssertEqual(result.source, .nativeFallback)
      XCTAssertEqual(result.status, expectedStatus)
    }

    let unknownFailure = SharedCoreProxyTypeRouter.route(
      rawType: "Synthetic",
      nativeClassification: proxy,
      classifyProxyType: { _ in throw SyntheticError() }
    )
    XCTAssertEqual(unknownFailure.classification, proxy)
    XCTAssertEqual(unknownFailure.source, .nativeFallback)
    XCTAssertEqual(unknownFailure.status, .unknownFailure)
  }

  func testShadowDeduplicatesSourceAndStatusAndIgnoresReporterFailure() {
    let observations = OSAllocatedUnfairLock(
      initialState: [SharedCoreProxyTypeShadowObservation]()
    )
    SharedCoreProxyTypeShadow.configure { observation in
      observations.withLock { $0.append(observation) }
      throw SyntheticError()
    }
    defer {
      SharedCoreProxyTypeShadow.configure(reporter: nil)
    }

    for _ in 0..<2 {
      XCTAssertEqual(
        SharedCoreProxyTypeShadow.observe(
          rawType: "Vmess",
          nativeClassification: proxy,
          classifyProxyType: { _ in .proxy }
        ),
        proxy
      )
    }
    XCTAssertEqual(
      SharedCoreProxyTypeShadow.observe(
        rawType: "Selector",
        nativeClassification: unknown,
        classifyProxyType: { _ in .unrecognized }
      ),
      unknown
    )

    XCTAssertEqual(
      observations.withLock { $0 },
      [
        SharedCoreProxyTypeShadowObservation(source: .sharedShadow, status: .matched),
        SharedCoreProxyTypeShadowObservation(source: .nativeFallback, status: .unrecognized),
      ]
    )
  }

  func testObservationGateKeysBySourceAndStatus() {
    var gate = SharedCoreProxyTypeShadowObservationGate()
    let matched = SharedCoreProxyTypeShadowObservation(
      source: .sharedShadow,
      status: .matched
    )
    let fallback = SharedCoreProxyTypeShadowObservation(
      source: .nativeFallback,
      status: .matched
    )

    XCTAssertTrue(gate.shouldReport(matched))
    XCTAssertFalse(gate.shouldReport(matched))
    XCTAssertTrue(gate.shouldReport(fallback))
    gate.reset()
    XCTAssertTrue(gate.shouldReport(matched))
  }

  private func route(
    sharedResult: SharedProxyTypeClassification,
    nativeClassification: ProxyClassification
  ) -> SharedCoreProxyTypeRouteResult {
    SharedCoreProxyTypeRouter.route(
      rawType: "Synthetic",
      nativeClassification: nativeClassification,
      classifyProxyType: { _ in sharedResult }
    )
  }

  private struct SyntheticError: Error {}
}
